### PR TITLE
Alerting: decouple api models from domain\dto models: separate Provenance status + converters

### PR DESCRIPTION
**Pull request by yuri-tceretian**
**What is this feature?**
This PR refactors Alerting API models (package "github.com/grafana/grafana/pkg/services/ngalert/api/tooling/definitions") and introduces a type Provenance that copies type Provenance from `github.com/grafana/grafana/pkg/services/ngalert/models`

Also it moves converters from and to API models to API package, new file `compat.go`

**Why do we need this feature?**
This is needed to decouple API models from the other packages so their imports will not affect the code generation.

**Which issue(s) does this PR fix?**:

Contributes to https://github.com/grafana/grafana/issues/63582

**Special notes for your reviewer**:

I chose to not implement explicit conversion between Provenance enums for now but I think there should be ones to enforce boundaries between API and domain layers

### CODE REVIEW

**Comment by alexmobo** [comment]
Would it be possible to use consistent naming throughout all packages when possible? I see `apimodels` in some files and `definitions` in others, both referring to the same package `"github.com/grafana/grafana/pkg/services/ngalert/api/tooling/definitions"`. It gets confusing when you are not entirely familiar with the code.

**Comment by yuri-tceretian** [comment]
> Would it be possible to use consistent naming throughout all files when possible? 

Good point! However, I would rather do this after we finish removing all dependencies from the definitions package. I would stick with `definitions`.

File: pkg/services/ngalert/provisioning/notification_policies_test.go [hunk 1]
```diff
@@ -158,7 +158,7 @@ func TestNotificationPolicyService(t *testing.T) {
 
 		updated, err := sut.GetPolicyTree(context.Background(), 1)
 		require.NoError(t, err)
-		require.Equal(t, models.ProvenanceAPI, updated.Provenance)
+		require.Equal(t, definitions.ProvenanceAPI, updated.Provenance)
```
Comment (alexweav) [inline pkg/services/ngalert/provisioning/notification_policies_test.go:161]: (just noting for readers)

Discussed for some time offline about this change and other `models->definitions` changes. The code being tested already works in terms of objects in `definitions`, and this ends up being OK in the long run.

File: pkg/services/ngalert/api/tooling/definitions/alertmanager.go [hunk 2]
```diff
@@ -567,11 +566,21 @@ func (c *PostableUserConfig) UnmarshalYAML(value *yaml.Node) error {
...
 }
 
+type Provenance string
+
+const (
+	// ProvenanceNone reflects the provenance when no provenance is stored
+	// for the requested object in the database.
+	ProvenanceNone Provenance = ""
+	ProvenanceAPI  Provenance = "api"
+	ProvenanceFile Provenance = "file"
```
Comment (alexweav) [inline pkg/services/ngalert/api/tooling/definitions/alertmanager.go:576]: Part of the principle we had was to keep the API layer's understanding of `provenance` more like a string than an enum. There was an original idea of supporting more provenance types, or allowing custom provenance types...

WDYT about doing something like `type Provenance string` alone in the API layer, and not specifying any actual constant values?

Most of the code in this layer cares about the type itself more than it cares about the values - and the conversion func can handle the matching.

File: pkg/services/ngalert/api/compat.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+	"github.com/prometheus/common/model"
+
+	"github.com/grafana/grafana/pkg/services/ngalert/api/tooling/definitions"
+	"github.com/grafana/grafana/pkg/services/ngalert/models"
+)
+
+// AlertRuleFromProvisionedAlertRule converts definitions.ProvisionedAlertRule to models.AlertRule
+func AlertRuleFromProvisionedAlertRule(a definitions.ProvisionedAlertRule) (models.AlertRule, error) {
+	return models.AlertRule{
```
Comment (alexweav) [inline pkg/services/ngalert/api/compat.go:14]: Just to start the discussion (not really feedback for this PR)

The biggest drawback of the conversion funcs is, it's really easy to write bugs if you forget to map a field.

I wonder how we can test the conversion funcs accurately both ways? In the past I've seen reflection be used for this - reflect into both types, generate some value, assert it gets mapped. Are there other approaches you've seen?

**Review by alexweav — APPROVED** [review]
LGTM, couple questions

File: pkg/services/ngalert/api/tooling/definitions/alertmanager.go (hunk 2 above)
Comment (yuri-tceretian) [reply pkg/services/ngalert/api/tooling/definitions/alertmanager.go:576]: +1 for that

File: pkg/services/ngalert/api/compat.go (hunk 3 above)
Comment (yuri-tceretian) [reply pkg/services/ngalert/api/compat.go:14]: That kind of problem is supposed to be caught by either integration or request handler unit tests.
A specific test for these functions can be solved by having a generator for models - https://github.com/brianvoe/gofakeit. I brought that backend architecture meeting last year, and there was not objections to adopting this library (iirc, It is used in K6).

File: pkg/services/ngalert/api/tooling/definitions/alertmanager.go (hunk 2 above)
Comment (yuri-tceretian) [reply pkg/services/ngalert/api/tooling/definitions/alertmanager.go]: https://github.com/grafana/grafana/pull/63594/commits/c6cb7d297e69a6ce4c3905b8711b3134d65964d2

**Pull request merged**